### ISSUE TITLE
Failed correctly deserialize Collections.unmodifiableMap(LinkedHashMap)

### ISSUE THREAD

**Issue by vivimice** [RuedigerMoeller/fast-serialization#128]
```java
Map<String, Integer> m1, m2;
		
m1 = new LinkedHashMap<String, Integer>();
m1.put("a", 1);
m1.put("BB", 1);
m1.put("ccc", 1);
m1 = Collections.unmodifiableMap(m1);
		
System.out.println(m1.keySet());
		
FSTConfiguration fstConf = FSTConfiguration.createDefaultConfiguration();
ByteArrayOutputStream out = new ByteArrayOutputStream();
fstConf.encodeToStream(out, m1);
m2 = (Map<String, Integer>) fstConf.decodeFromStream(new ByteArrayInputStream(out.toByteArray()));

System.out.println(m2.keySet());
```

Output is:
```
[a, BB, ccc]
[BB, a, ccc]
```

FSTVersion: 2.45

**Comment by vivimice** [comment]
It seems that 2.45 will break Collections.unmodifiableSet(LinkedHashSet) too

**Comment by vivimice** [comment]
Since Collection.unmodifiableSet/unmodifiableMap is UNMODIFIABLE, so use LinkedHashSet/LinkedHashMap as inner Set/Map is reasonable. (No matter what type of Map/Set is wrapped within the unmodifiable instance, the iteration order will be as same as serialization order)

I think the following patches will work:

```
--- FSTUnmodifiableMapSerializer.java   2016-05-10 14:30:52.574374357 +0800
+++ FSTUnmodifiableMapSerializer.fixed.java     2016-05-10 14:31:31.964065611 +0800
@@ -18,7 +18,7 @@
 import org.nustaq.serialization.util.FSTUtil;
 
 import java.util.Collections;
-import java.util.HashMap;
+import java.util.LinkedHashMap;
 import java.util.Map;
 
 /**
@@ -42,7 +42,7 @@
         try {
             int len = in.readInt();
             if (UNMODIFIABLE_MAP_CLASS.isAssignableFrom(objectClass)) {
-                Map res = new HashMap(len);
+                Map res = new LinkedHashMap(len);
                 in.registerObject(res, streamPosition, serializationInfo, referencee);
                 for (int i = 0; i < len; i++) {
                     Object key = in.readObjectInternal(null);
```

```
--- FSTUnmodifiableCollectionSerializer.java    2016-05-10 14:32:45.561521297 +0800
+++ FSTUnmodifiableCollectionSerializer.fixed.java      2016-05-10 14:33:15.812564487 +0800
@@ -21,7 +21,7 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.HashSet;
+import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Set;
 
@@ -56,7 +56,7 @@
                 return Collections.unmodifiableList(res);
             }
             if (UNMODIFIABLE_SET_CLASS.isAssignableFrom(objectClass)) {
-                Set res = new HashSet(len);
+                Set res = new LinkedHashSet(len);
                 fillArray(in, serializationInfo, referencee, streamPosition, res, len);
                 return Collections.unmodifiableSet(res);
             }
```

**Comment by RuedigerMoeller** [comment]
thanks a lot for your input, I'll incorporate these into 2.46

**Comment by johnou** [comment]
@RuedigerMoeller do you have a plan on when 2.46 will be released with this patch?

**Comment by jkubrynski** [comment]
Are you sure it's really a case? When you define field as a Map it means you don't care about order. When you care about order you use SortedMap. The same applies to Set. Set by general means order is not important. You use SortedSet when you need it. I don't think using LinkedHashMap everywhere is a good idea.

**Comment by johnou** [comment]
Yep, we have APIs eg. scoreboards which return a set and the order needs to be maintained but to prevent accidental mutation we wrap it in an unmodifiableSet.

**Comment by jkubrynski** [comment]
@johnou then IMHO you should define this field as a SortedSet instead of Set. Otherwise you expect from Set something that isn't its responsitibilty

**Comment by mafagafogigante** [comment]
Aren't there unmodifiableSortedMap and unmodifiableSortedSet?

**Comment by johnou** [comment]
@mafagafogigante wow, missed that guy. I'll update the pull request in an hour or so. Thanks for the feedback guys!

**Comment by johnou** [comment]
still on the fence here, @jkubrynski javadocs [1] state the following;

This implementation spares its clients from the unspecified, generally chaotic ordering provided by HashSet, without incurring the increased cost associated with TreeSet. It can be used to produce a copy of a set that has the same order as the original, regardless of the original set's implementation:

     void foo(Set s) {
         Set copy = new LinkedHashSet(s);
         ...
     }
 
This technique is particularly useful if a module takes a set on input, copies it, and later returns results whose order is determined by that of the copy. (Clients generally appreciate having things returned in the same order they were presented.)

[1] http://docs.oracle.com/javase/7/docs/api/java/util/LinkedHashSet.html

**Comment by jkubrynski** [comment]
Let the @RuedigerMoeller decide. I think using LinkedHashSet everytime someone uses Set means solving problem which doesn't exists.

**Comment by mafagafogigante** [comment]
I agree with @jkubrynski. Set does not require ordering and performance really does matter on a library such as this one (which is already being used to serialize / deserialize GBs of objects). The unmodifiableSorted* stuff **needs** preserved ordering. Not preserving ordering there would be a bug. A simple unmodifiableSet does not. If this was free - performance and memory-wise - I would support you. But it is not.

**Comment by vivimice** [comment]
I agree with @jkubrynski too. Using LinkedHash* everywhere is a bad idea if the programmer don't known what he/she exactly want to do. 

Since Collections.unmodifiable* is just a view of the wrapped collection/map object, the best (and theoretical) way of implementing this deserializer is to read the data into the same original collection/map class. 

I've considerered this, but it is not easy: Take unmodifiedSet for example, there may be hundreds of implementations of java.util.Set, its impossible (or way too complicated) for the deserializer to instance the correspoding class. So using LinkedHashSet is just a trade-off: It is a Set, and preserves the iteration order - no matter the order is sorted or by-access or by-insert or even undefined in the orginal set, and, last but no the least, the implementation is simple.

**Comment by RuedigerMoeller** [comment]
Hi, in principle, the serialization contract states that deserialized data should be same as serialized one, regardless of variable typing. So if I specify a field of type "Set" this does not imply a serializer is free to arbitrary change the Set implementation of that fields value upon deserialization.

Haven't had a detailed look, but in general it the unmodifiable wrapped set implementation should be serializable and therefore the correct set implementation should be instantiated on deserialization. There might be tradeoffs required caused to nasty detail problems (instantiating classes with private default constructors etc.). I'll have a closer look this week and will release 2.46 next week. Unfortunately I don't have time to retest the Android N related fixes / PR's, so I'll have to release them blindly.

**Comment by johnou** [comment]
@RuedigerMoeller when reading the deserialization data is fast-serialization able to determine the internal collection type of the wrapped collection?

**Comment by RuedigerMoeller** [comment]
if its written correctly (aka out.writeObject() etc.), yes. Checkout FSTCollectionSerializer writing its elements (you can use the default JDK API out.writeObject() to write an arbitrary serializable from a FSTSerializer impl)

**Comment by johnou** [comment]
Kryo uses reflection to access the wrapped set / map impl, writes a marker for what type of unmodifiable collection then writes the wrapped set / map object.

https://github.com/magro/kryo-serializers/blob/c1e6bb3c849adbabeea36fbea0670fd48656f360/src/main/java/de/javakaffee/kryoserializers/UnmodifiableCollectionsSerializer.java

**Comment by RuedigerMoeller** [comment]
Hi, i tried to reproduce the original issue (having removed all pull request since) and could not reproduce the intial error.

As all unmodifiable* are serializable, FST correctly then serializes the inner collection LinkedHashMap using FSTMapSerializer which in turn puts the elements exactly in the order as written.

```java
    @Test
    public void testUnmodifiableLinkedHashMap() {
        FSTConfiguration conf = FSTConfiguration.createDefaultConfiguration();

        Map m1 = new LinkedHashMap<String, Integer>();
        m1.put("a", 1);
        m1.put("BB", 1);
        m1.put("ccc", 1);
        m1.put("aa", 1);
        m1.put("BBa", 1);
        m1.put("ccca", 1);
        m1 = Collections.unmodifiableMap(m1);

        Map un = Collections.unmodifiableMap(m1);
        Map res = (Map) conf.asObject(conf.asByteArray(un));
        assertTrue(DeepEquals.deepEquals(res,un));

        System.out.println(res.keySet());
        System.out.println(m1.keySet());
    }
```
result in:

```
[a, BB, ccc, aa, BBa, ccca]
[a, BB, ccc, aa, BBa, ccca]
```

so as the linkedhashmap is written ordered using FSTMapSerializer

```java
    @Override
    public void writeObject(FSTObjectOutput out, Object toWrite, FSTClazzInfo clzInfo, FSTClazzInfo.FSTFieldInfo referencedBy, int streamPosition) throws IOException {
        Map col = (Map)toWrite;
        out.writeInt(col.size());
        FSTClazzInfo lastKClzI = null;
        FSTClazzInfo lastVClzI = null;
        Class lastKClz = null;
        Class lastVClz = null;
        for (Iterator iterator = col.entrySet().iterator(); iterator.hasNext(); ) {
            Map.Entry next = (Map.Entry) iterator.next();
            Object key = next.getKey();
            Object value = next.getValue();
            if ( key != null && value != null ) {
                lastKClzI = out.writeObjectInternal(key, key.getClass() == lastKClz ? lastKClzI : null, null);
                lastVClzI = out.writeObjectInternal(value, value.getClass() == lastVClz ? lastVClzI : null, null);
                lastKClz = key.getClass();
                lastVClz = value.getClass();
            } else
            {
                out.writeObjectInternal(key, null, null);
                out.writeObjectInternal(value, null, null);
            }

        }
    }
```
and on read time (also FSTMapSerializer)

```java
        Object res = null;
        int len = in.readInt();
        if ( objectClass == HashMap.class ) {
            res = new HashMap(len);
        } else
        if ( objectClass == Hashtable.class ) {
            res = new Hashtable(len);
        } else
        {
            res = objectClass.newInstance();
        }
        in.registerObject(res, streamPosition,serializationInfo, referencee);
        Map col = (Map)res;
        for ( int i = 0; i < len; i++ ) {
            Object key = in.readObjectInternal(null);
            Object val = in.readObjectInternal(null);
            col.put(key,val);
        }
        return res;
```

the map is reconstructed in exactly the same order (the testcase also does not fail with my 1.8_45). I think this is superior to explicit handling as everything is handled by FST falling back to generic default JDK-emulation.

as FSTConfiguration registers not for subclasses of LinkedHashMap
```java
        reg.putSerializer(LinkedHashMap.class, new FSTMapSerializer(), false); // subclass should register manually
```

could it be you where serializing a subclass of LinkedHashMap ? The problem in my view has nothing to do with unmodifiable, as Collections.unmodifiable* just wraps the existing collection and FST just serializes the wrapper and its contained collection as usual.

Unsure. Really need a reproducer (which platform/JDK did you use ?)

**Comment by RuedigerMoeller** [comment]
another note (from pull requests)

```java
public Object instantiate(Class objectClass, FSTObjectInput in, FSTClazzInfo serializationInfo, FSTClazzInfo.FSTFieldInfo referencee, int streamPosition) throws Exception {
        try {
            int len = in.readInt();
            if (UNMODIFIABLE_MAP_CLASS.isAssignableFrom(objectClass)) {
                Map res = new LinkedHashMap(len);
                in.registerObject(res, streamPosition, serializationInfo, referencee);
                for (int i = 0; i < len; i++) {
                    Object key = in.readObjectInternal(null);
                    Object val = in.readObjectInternal(null);
                    res.put(key, val);
                }
                return Collections.unmodifiableMap(res);
            }
        } catch (Throwable th) {
            FSTUtil.<RuntimeException>rethrow(th);
        }
        return null;
    }
```
turning any unmodifiable wrapped Map into a LinkedHashMap is a no-go as it breaks the serialization contract. Better register this serializer especially for your application, as a default behaviour it breaks things ;) .

**Comment by RuedigerMoeller** [comment]
seems this came from #114 which is only JSON related. The Serializers introduced should be used for JSON configuration only ...

**Comment by vivimice** [comment]
@RuedigerMoeller The test case you gave produces different output on my machine.

```
[BB, aa, BBa, ccca, a, ccc]
[a, BB, ccc, aa, BBa, ccca]
```

Note: I commented out the ```assertTrue(...)``` line

I created a fresh project to reproduce this:

JDK: jdk1.8.0_40 (Oracle)
Platform: Windows 10 Pro Insider Preview (64bit)
Libraries and md5sums:
```
e2aa4073a4ad1c0a8082e2c6781e39aa  fst-2.45.jar
5763215f9054614336ffa55d46a69e15  jackson-core-2.5.3.jar
32ccb1d20a42b5aaaceb90c9082a2efa  objenesis-2.1.jar
```

Jars downloaded from mvnrepository.com
```
http://central.maven.org/maven2/de/ruedigermoeller/fst/2.45/fst-2.45.jar
http://central.maven.org/maven2/com/fasterxml/jackson/core/jackson-core/2.5.3/jackson-core-2.5.3.jar
http://central.maven.org/maven2/org/objenesis/objenesis/2.1/objenesis-2.1.jar
```

Note: also produced the same wrong answer on Debian 8 with jdk:
```
java version "1.8.0_74"
Java(TM) SE Runtime Environment (build 1.8.0_74-b02)
Java HotSpot(TM) 64-Bit Server VM (build 25.74-b02, mixed mode)
```

See the attachment for the reproducing project.

[fst-unmodifiable-reproduce.tar.gz](https://github.com/RuedigerMoeller/fast-serialization/files/276698/fst-unmodifiable-reproduce.tar.gz)

**Comment by RuedigerMoeller** [comment]
Yes, I tested removing changes of a pull request meant to be applied to JSON codec only which actually introduced a bunch of issues for default serialization.

Will deliver a fix ASAP.

**Comment by johnou** [comment]
@RuedigerMoeller so I did a little more digging and I found the unmodifiable serializers are not completely necessary, infact there seems to be a misconfiguration with the map serializers, subclasses is set as true but there is a java comment saying subclasses should be defined individually.. When using a json configuration (compatibility mode and without the unmodifiable serializers) the only test which blows up is the one trying to serialize java.util.Collections$UnmodifiableRandomAccessList.

You can see my test branch here https://github.com/johnou/fast-serialization/commit/b1baba9a293eb331d933783b44409de80773fc74

**Comment by RuedigerMoeller** [comment]
unmodifiable serializers are only required for JSON, as JSON codec does not support JDK-serialization-emulation. I did some adaptions to the pull request's and released them with 2.46.
Thanks everyone for detecting and helping out :)
